### PR TITLE
Check if node has saved_output_filename

### LINKED ISSUE #5: Add CI for linting, formatting, and tests

### PR DESCRIPTION
**Pull request by sjawhar**
I ran into this while doing some testing for vivaria:
```
Traceback (most recent call last):
  File "/home/agent/.local/lib/python3.11/site-packages/pyhooks/__init__.py", line 266, in error_handler_wrapper
    await main_function(self)
  File "/home/agent/.agent_code/main.py", line 153, in main
    await getattr(prompters, agent.settings.prompter)(agent)
  File "/home/agent/.agent_code/modules/prompters.py", line 160, in _context_and_usage_aware
    _get_trimmed_message(agent.state.nodes[node_id], token_usage_fraction)
  File "/home/agent/.agent_code/modules/prompters.py", line 141, in _get_trimmed_message
    filename=node.metadata["saved_output_filename"],
             ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 'saved_output_filename'
```

And then while fixing it I decided it was high time we had some good CI on this repo 😸 

Closes #5

### CODE REVIEW

File: requirements.txt
Comment (sjawhar) [inline requirements.txt:1]: This file will be automatically generated by Poetry on CI. Full locking for all dependencies

File: pyproject.toml
Comment (sjawhar) [inline pyproject.toml:16]: Can update these versions in #6 after merging this PR (I can help)

File: modules/discriminators.py
Comment (sjawhar) [inline modules/discriminators.py:142]: Define the variable before starting the `for` loop, so it's defined even if the loop is of size 0

File: modules/discriminators.py
Comment (sjawhar) [inline modules/discriminators.py:185]: Should an error be raised here instead, if `generations.outputs is None`?

File: modules/discriminators.py
Comment (sjawhar) [inline modules/discriminators.py:254]: Is this OK?

File: modules/generators.py
Comment (sjawhar) [inline modules/generators.py:69]: Is this OK?

File: modules/generators.py
Comment (sjawhar) [inline modules/generators.py:77]: Simpler way to do the thing, no type errors

File: modules/generators.py
Comment (sjawhar) [inline modules/generators.py:167]: Is this OK?

File: modules/prompters.py
Comment (sjawhar) [inline modules/prompters.py:145]: This was the original bug

Comment (CosmicGans) [reply modules/prompters.py:145]: Thanks for looking into this! 
I am confused however: why are there sometimes truncated outputs, but no saved_output_filenames?

Comment (sjawhar) [reply modules/prompters.py:145]: `saved_output_filenames` and truncated outputs are separate things. This code here does the truncation, whereas saving output to a file is done elsewhere. This code is currently assuming that outputs of commands are always saved to a file, which is not true (it depends on the actor).

**Comment by sjawhar** [comment]
Bump

File: modules/discriminators.py
Comment (satojk) [reply modules/discriminators.py:185]: I think it's plausible for there to be no outputs and for us to want to try again, so seems correct to not raise an error.

File: modules/discriminators.py
Comment (satojk) [reply modules/discriminators.py:254]: Seems better than what's currently there which will also just error out, but I think ideally we retry the generation.

File: modules/discriminators.py
Comment (satojk) [inline modules/discriminators.py:77]: Hm, this makes me think that we shouldn't be defining the Message class in base.py, which is just a stronger-typed version of OpenaiChatMessage.

File: modules/discriminators.py
Comment (satojk) [reply modules/discriminators.py:254]: Or actually I think my assumption is that this never fails and in fact hasn't failed. So I wouldn't want to just silently retry. If it fails then something is up. Seems good to explicitly raise the error.

File: modules/generators.py
Comment (satojk) [reply modules/generators.py:167]: IIUC this would never trigger because you can only break out of the loop above if generations is not None. In either case if you've successfully broken out of the loop, `good_generations` is what you really care about, so I think we can remove this.

File: modules/tools.py
Comment (satojk) [inline modules/tools.py:105]: Is it better to just make `submission` a kwarg and set `""` as the default?

File: templates.py
Comment (satojk) [inline templates.py:57]: Nice!

**Review by satojk — APPROVED** [review]

File: modules/tools.py
Comment (sjawhar) [reply modules/tools.py:105]: Do you mean as the argument to `double_return_fn`? That feels like a riskier change, I'm not confident that the function never gets called with `submission=None`

File: modules/generators.py
Comment (sjawhar) [reply modules/generators.py:167]: `generations` is called below on line 180: `generations.dict().items()`

File: templates.py
Comment (sjawhar) [reply templates.py]: Noticed the message still referred to a file in cases where it might not exist. Updated

File: modules/tools.py
Comment (satojk) [reply modules/tools.py:105]: Yeah, I was thinking that'd be fine (using `""` whenever this gets called with `submission=None`, if ever). At any rate we probably never want to call `hooks.submit(None)` (I think this causes an error), which could still happen on line 99. 

(also I just noticed that it's already a kwarg with default `None`, so I guess my suggestion is to just change the default from `None` to `""`).

File: .github/workflows/pr-and-main.yaml
Comment (satojk) [inline .github/workflows/pr-and-main.yaml:70]: Oh this is great!

File: modules/tools.py
Comment (sjawhar) [reply modules/tools.py:105]: > we probably never want to call `hooks.submit(None)` (I think this causes an error), which could still happen on line 99.

No, it shouldn't, because `if submission in _state.submissions` means `submission` is a string because `_state.submissions` is a list of strings.

> I just noticed that it's already a kwarg with default None, so I guess my suggestion is to just change the default from `None` to `""`

That doesn't handle the case where the function gets called with `submission=None`, which I don't know to never happen. Note that this PR is not changing the signature of the function.

**Comment by neevparikh** [comment]
@sjawhar any reason we're still waiting on this branch? #6 depends on this for the checks to pass I think, or at least @satojk is waiting for this to merge.

**Pull request merged**
